misc/peer-store/src/behaviour.rs

```diff
@@ -0,0 +1,122 @@
+use std::{collections::VecDeque, task::Poll};
+
+use libp2p_core::{Multiaddr, PeerId};
+use libp2p_swarm::{dummy, FromSwarm, NetworkBehaviour};
+
+use crate::store::Store;
+
+/// Events of this behaviour that will be emmitted to the swarm.
+pub enum Event {
+    RecordUpdated { peer: PeerId },
+}
+
+pub struct Behaviour<S> {
+    store: S,
+    /// Events that will be emitted.
+    pending_events: VecDeque<Event>,
+}
+
+impl<S> Behaviour<S>
+where
+    S: Store + 'static,
+{
+    pub fn new(store: S) -> Self {
+        Self {
+            store,
+            pending_events: VecDeque::new(),
+        }
+    }
+    /// List peers that are currently connected to this peer.
+    pub fn list_connected(&self) -> Box<[&PeerId]> {
+        self.store.list_connected()
+    }
+    /// Try to get all observed address of the given peer.  
+    /// Returns `None` when the peer is not in the store.
+    pub fn address_of_peer(&self, peer: &PeerId) -> Option<Box<[super::AddressRecord]>> {
+        self.store.addresses_of_peer(peer)
+    }
+    /// Manually update a record.  
+    /// This will always cause an `Event::RecordUpdated` to be emitted.
+    pub fn update_record(&mut self, peer: &PeerId, address: &Multiaddr) {
+        self.store.on_address_update(peer, address);
+        self.pending_events
+            .push_back(Event::RecordUpdated { peer: *peer });
+    }
+}
+
+impl<S> NetworkBehaviour for Behaviour<S>

```

Comment:
The behavior must implement `handle_pending_outbound_connection`, so it can provide the known addresses to the swarm when attempting to dial a peer.